src/compute/sort/row/mod.rs

```diff
@@ -0,0 +1,1041 @@
+    pub fn new_with_options(data_type: DataType, options: SortOptions) -> Self {
+        Self { options, data_type }
+    }
+}
+
+impl RowConverter {
+    /// Create a new [`RowConverter`] with the provided schema
+    pub fn new(fields: Vec<SortField>) -> Self {
+        let interners = (0..fields.len()).map(|_| None).collect();
+        Self {
+            fields: fields.into(),
+            interners,
+        }
+    }
+
+    /// Convert [`ArrayRef`] columns into [`Rows`]
+    ///
+    /// See [`Row`] for information on when [`Row`] can be compared
+    ///
+    /// # Panics
+    ///
+    /// Panics if the schema of `columns` does not match that provided to [`RowConverter::new`]
+    pub fn convert_columns(&mut self, columns: &[Box<dyn Array>]) -> Result<Rows> {
+        if columns.len() != self.fields.len() {
+            return Err(Error::InvalidArgumentError(format!(
+                "Incorrect number of arrays provided to RowConverter, expected {} got {}",
+                self.fields.len(),
+                columns.len()
+            )));
+        }
+
+        let dictionaries = columns
+            .iter()
+            .zip(&mut self.interners)
+            .zip(self.fields.iter())
+            .map(|((column, interner), field)| {
+                if column.data_type() != &field.data_type {
+                    return Err(Error::InvalidArgumentError(format!(
+                        "RowConverter column schema mismatch, expected {:?} got {:?}",
+                        field.data_type,
+                        column.data_type()
+                    )));
+                }
+
+                let values = match column.data_type() {
+                    DataType::Dictionary(k, _, _) => match_integer_type!(k, |$T| {
+                        let column = column
+                            .as_any()
+                            .downcast_ref::<DictionaryArray<$T>>()
+                            .unwrap();
+                        column.values()
+                    }),
+                    _ => return Ok(None),
+                };
+
+                let interner = interner.get_or_insert_with(Default::default);
+
+                let mapping = compute_dictionary_mapping(interner, values)?
+                    .into_iter()
+                    .map(|maybe_interned| {
+                        maybe_interned.map(|interned| interner.normalized_key(interned))
+                    })
+                    .collect::<Vec<_>>();
+
+                Ok(Some(mapping))
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        let mut rows = new_empty_rows(columns, &dictionaries, Arc::clone(&self.fields))?;
+
+        for ((column, field), dictionary) in
+            columns.iter().zip(self.fields.iter()).zip(dictionaries)
+        {
+            // We encode a column at a time to minimise dispatch overheads
+            encode_column(&mut rows, column, field.options, dictionary.as_deref())
```

Comment:
This is almost parallelizable - it is changing `rows`. However, there is still an optimization since modifying `rows` is `O(1)` but encoding is `O(C)`. Will continue to think about this.

Reply:
I'm puzzled about how to convert this into a parallel one?

Reply:
I agree.

The idea would be that we can build `Vec<u8>` per column in parallel and only write them to the rows afterwards. However, I realized later that "encoding" is very simple and low overhead, so I am not sure it is worth.